synchronizer/synchronizer.go

```diff
@@ -682,3 +730,113 @@ func (s *ClientSynchronizer) processVerifiedBatch(verifiedBatch etherman.Verifie
+
+		if matchNumber && matchGER && matchLER && matchSR &&
+			matchCoinbase && matchTimestamp && matchL2Data {
+			log.Debugf("batch %v already synchronized", trustedBatch.BatchNumber)
+			return nil
+		}
+		log.Infof("batch %v needs to be updated", trustedBatch.BatchNumber)
+	} else {
+		log.Infof("batch %v needs to be synchronized", trustedBatch.BatchNumber)
+	}
+
+	log.Debugf("resetting trusted state from batch %v", trustedBatch.BatchNumber)
+	if err := s.state.ResetTrustedState(s.ctx, trustedBatch.BatchNumber-1, dbTx); err != nil {
+		log.Errorf("failed to reset trusted state", trustedBatch.BatchNumber)
+		return err
+	}
+
+	log.Debugf("opening batch %v", trustedBatch.BatchNumber)
+	if err := s.state.OpenBatch(s.ctx, processCtx, dbTx); err != nil {
+		log.Errorf("error opening batch %d", trustedBatch.BatchNumber)
+		return err
+	}
+
+	log.Debugf("processing sequencer for batch %v", trustedBatch.BatchNumber)
+	processBatchResp, err := s.state.ProcessSequencerBatch(s.ctx, trustedBatch.BatchNumber, txs, dbTx)
+	if err != nil {
+		log.Errorf("error processing sequencer batch for batch: %d", trustedBatch.BatchNumber)
+		return err
+	}
+
+	log.Debugf("storing transactions for batch %v", trustedBatch.BatchNumber)
+	if err = s.state.StoreTransactions(s.ctx, trustedBatch.BatchNumber, processBatchResp.Responses, dbTx); err != nil {
+		log.Errorf("failed to store transactions for batch: %d", trustedBatch.BatchNumber)
+		return err
+	}
+
+	log.Debug("trustedBatch.StateRoot ", trustedBatch.StateRoot)
+	isBatchClosed := trustedBatch.StateRoot != state.ZeroHash.String()
+	if isBatchClosed {
+		receipt := state.ProcessingReceipt{
+			BatchNumber:   trustedBatch.BatchNumber,
+			StateRoot:     common.HexToHash(trustedBatch.StateRoot),
+			LocalExitRoot: common.HexToHash(trustedBatch.LocalExitRoot),
+		}
+		log.Debugf("closing batch %v", trustedBatch.BatchNumber)
+		if err := s.state.CloseBatch(s.ctx, receipt, dbTx); err != nil {
+			log.Errorf("error closing batch %d", trustedBatch.BatchNumber)
+			return err
+		}
+	}
+
+	if trustedBatch.ForcedBatchNumber > 0 {
+		log.Debugf("batch %v was forced with number forced batch number %v", trustedBatch.BatchNumber, trustedBatch.ForcedBatchNumber)
+		forcedBatch := &state.ForcedBatch{
+			// BlockNumber:       trustedBatch.,
+			BatchNumber:       &trustedBatch.BatchNumber,
+			ForcedBatchNumber: trustedBatch.ForcedBatchNumber,
+			Sequencer:         common.HexToAddress(trustedBatch.Sequencer),
+			GlobalExitRoot:    common.HexToHash(trustedBatch.GlobalExitRoot),
+			RawTxsData:        trustedBatchL2Data,
+			// ForcedAt:          trustedBatch.,
+		}
+
+		log.Debugf("adding forced batch %v for batch %v", trustedBatch.ForcedBatchNumber, trustedBatch.BatchNumber)
+		if err := s.state.AddForcedBatch(s.ctx, forcedBatch, dbTx); err != nil {
+			log.Errorf("error adding forced batch for batch %v", trustedBatch.BatchNumber)
+			return err
+		}
```

Comment:
I guess that here you are adding the batchNumber field to the forcedBatch to remove it from the queue and leave it in a stated "used". Am I right?

Reply:
No, `AddForcedBatch` is inserting a new batch. What should happen here is what @ARR552 suggests:

- The forced batch should already exist on the DB
- We need to mark this batch as sequenced (the forced batch has moved from the forced queue into the state). This is done by updating the `batch_num` from `NULL` to `trustedBatch.BatchNumber`